src/Nethermind/Nethermind.Merge.AuRa/Shutter/ShutterP2P.cs

```diff
@@ -0,0 +1,131 @@
+using System.Collections.Generic;
+using Multiformats.Address;
+using Nethermind.Consensus.AuRa.Config;
+using Nethermind.Logging;
+using ILogger = Nethermind.Logging.ILogger;
+using System.Threading.Channels;
+
+namespace Nethermind.Merge.AuRa.Shutter;
+
+public class ShutterP2P(
+    Action<Dto.DecryptionKeys> onDecryptionKeysReceived,
+    IAuraConfig auraConfig,
+    ILogManager logManager)
+{
+    private readonly ILogger _logger = logManager.GetClassLogger();
+    private readonly Channel<byte[]> _msgQueue = Channel.CreateUnbounded<byte[]>();
+    private PubsubRouter? _router;
+    private ServiceProvider? _serviceProvider;
+    private CancellationTokenSource? _cancellationTokenSource;
+    private const int DisconnectionLogTimeout = 60 * 5;
+
+    public class ShutterP2PException(string message, Exception? innerException = null) : Exception(message, innerException);
+
+    public void Start(in IEnumerable<string> p2pAddresses)
+    {
+        _serviceProvider = new ServiceCollection()
+            .AddLibp2p(builder => builder)
+            .AddSingleton(new IdentifyProtocolSettings
+            {
+                ProtocolVersion = auraConfig.ShutterP2PProtocolVersion,
+                AgentVersion = auraConfig.ShutterP2PAgentVersion
+            })
+            .AddSingleton(new Settings()
+            {
+                ReconnectionAttempts = int.MaxValue
+            })
+            // .AddLogging(builder =>
+            //     builder.SetMinimumLevel(Microsoft.Extensions.Logging.LogLevel.Trace)
+            //     .AddSimpleConsole(l =>
+            //     {
+            //         l.SingleLine = true;
+            //         l.TimestampFormat = "[HH:mm:ss.FFF]";
+            //     })
+            // )
+            .BuildServiceProvider();
+
+        IPeerFactory peerFactory = _serviceProvider.GetService<IPeerFactory>()!;
+        ILocalPeer peer = peerFactory.Create(new Identity(), "/ip4/0.0.0.0/tcp/" + auraConfig.ShutterP2PPort);
+        if (_logger.IsInfo) _logger.Info($"Started Shutter P2P: {peer.Address}");
+        _router = _serviceProvider.GetService<PubsubRouter>()!;
+
+        ITopic topic = _router.Subscribe("decryptionKeys");
+
+        topic.OnMessage += (byte[] msg) =>
+        {
+            _msgQueue.Writer.TryWrite(msg);
+            if (_logger.IsDebug) _logger.Debug("Received Shutter P2P message.");
+        };
+
+        MyProto proto = new();
+        _cancellationTokenSource = new();
+        _ = _router.RunAsync(peer, proto, token: _cancellationTokenSource.Token);
+        ConnectToPeers(proto, p2pAddresses);
+
+        Task.Run(async () =>
+        {
+            while (true)
+            {
+                var msg = await _msgQueue.Reader.ReadAsync();
+                try
+                {
+                    ProcessP2PMessage(msg);
+                }
+                catch (Exception e)
+                {
+                    throw new ShutterP2PException("Shutter processing thread error", e);
```

Comment:
It would be better if the thread didn't die, hopefully exceptions should be caught before this point. I can use a bounded channel instead to prevent this memory leak